src/ConfigFileParser.cpp

```diff
@@ -0,0 +1,534 @@
+#include "ConfigFileParser.hpp"
+
+/**
+ * @brief PUBLIC Construct a new ConfigFileParser:: ConfigFileParser object.
+ */
+ConfigFileParser::ConfigFileParser(void) { }
+
+/**
+ * @brief PRIVATE Construct a new ConfigFileParser:: ConfigFileParser object with copy.
+ *
+ * The copy constructor is private to prevent direct object creation.
+ * @param ref The ConfigFileParser object to copy
+ */
+ConfigFileParser::ConfigFileParser(const ConfigFileParser& ref)
+{
+    static_cast<void>(ref);
+}
+
+/**
+ * @brief PRIVATE Overload of the assignment operator.
+ *
+ * The assigment operator is private to prevent direct object creation.
+ * @param ref The ConfigFileParser object to assign.
+ * @return ConfigFileParser&  The ConfigFileParser object.
+ */
+ConfigFileParser& ConfigFileParser::operator=(const ConfigFileParser& ref)
+{
+    static_cast<void>(ref);
+    return *this;
+}
+
+/**
+ * @brief Parses the config file into a ConfigFile object and returns this object
+ *
+ * The function checks also if the config file:
+ * 1. Can be opened
+ * 2. Is not empty
+ * 3. Does not contain open brackets
+ * 4. Starts with 'http {'
+ * 5. Contains minimum one server
+ *
+ * @param configFilePath Path to the config file
+ * @return const ConfigFile& Created ConfigFile object
+ */
+const ConfigFile& ConfigFileParser::parseConfigFile(const std::string& configFilePath)
+{
+    m_configFile.stream.open(configFilePath.c_str());
+    if (!m_configFile.stream.is_open())
+        throw std::runtime_error("Failed to open config file");
+    if (m_configFile.stream.peek() == std::ifstream::traits_type::eof())
+        throw std::runtime_error("Config file is empty");
+
+    if (isBracketOpen(configFilePath))
+        throw std::runtime_error("Open bracket(s) in config file");
+
+    readAndTrimLine();
+    if (m_configFile.currentLine != "http {")
+        throw std::runtime_error("Config file does not start with 'http {'");
+
+    m_configFile.serverIndex = 0;
+    for (readAndTrimLine(); m_configFile.currentLine != "}"; readAndTrimLine()) {
```

Comment:
I think this is a for loop like `for (;  m_configFile.currentLine != "}"; ) { `
which could be just a while loop?
It looks like readAndTrimLine() are also doing stuff for the for loop > I would move then into the loop to make more readable

Reply:
For me this is an example of "smart" code: is complex for no real upside other than being quite elegant.

I know how it works. But I have to think about it. Why?

Reply:
But isn't the thinking due to the fact that we haven't seen this technique before and needed to understand it for the first time? When we see it more often, we get used to it and it doesn't require additional "brain power". Take for example: var1 += var2. When you don't know the concept, it may also seem confusing. But now it is clear as day to us because we have seen and used it so many times.

Reply:
I think this is a question of syntax vs idiom.

For me the difference is that `var1 += var2` is an idiom of the language. I think you meant to say it means take var2 and add it to var1. You learn it, you know it. And since almost everybody uses it the same way you know what to expect.

But the syntax in C++ would be something like `Num operator+=(const Num& rhs)`
You can also use the += operator to minus something (maybe it needs to do that for whatever case)
Or how I use it in Logger, where if you Logger += LogData: the Data doesn't get "added" per se but it gets logged, which I thought is something similar as adding

The for-loop syntax is `for (statement1; statement2; statement3) { code_block }`. 
But I'd argue that the idiomatic for-loop is `for (variable decl; condition; increment) { code_block }`.
Whereas `for (do_stuff_once(); condition; do_stuff_more() );` is (only) syntactically correct.

So the question becomes: why break the idiom (which everybody knows and uses)? 
If it would have some other upside I would go for it! But like I said I think in this case its just clever code.

I don't think we would see it more often.

Reply:
Only time can tell...

Reply:
Since we didn't see it yet, I'd argue it's not idiomatic.

When time tells we can switch

Reply:
What's your take on this @QCHR1581 ? At the end it is your code :D

Reply:
I would like to leave it like this. But if I rewrite the parsing this topic could solve itself, because I would probably choose another approach.

Reply:
Ok fine for me